packages/hardhat-core/src/internal/vendor/await-semaphore/index.ts

```diff
@@ -64,3 +67,154 @@ export class Mutex extends Semaphore {
     super(1);
   }
 }
+
+export class MultiProcessMutex {
+  // Logic explanation: the fs.writeFile function, when used with the wx+ flag, performs an atomic operation to create a file.
+  // If multiple processes try to create the same file simultaneously, only one will succeed.
+  // This logic can be utilized to implement a mutex.
+  // For more info check the Nomic Notion page (internal link).
+  // Timeout logic explanation:
+  // When the function passed to the mutex is executed, it must not take longer than _defaultTimeoutInMs.
+  // If it exceeds this duration, the function is terminated, and the mutex is released.
+  // In case the process running the function crashes, other processes waiting for the mutex can detect this
+  // by checking the elapsed time since the mutex was acquired. This approach helps to avoid deadlock scenarios.
```

Comment:
Pedantic comment, but: AFAIK that's not a deadlock. It's just a stale lock file.